### ISSUE TITLE
should have any way to verify sent/recived/downloaded/uploaded files

### ISSUE THREAD

**Issue by mzramna** [LonamiWebs/Telethon#1647]
sould have an way to verify the checksum of a file before it be downloaded and after been uploaded.
this could be used in many scenarios,like verify if a file has been sent to a chat or if this apears multiple times or even if its corrupted,other utility is to verify if a downloaded media already exists inside the download folder or anywhere else the software checks
if it already exists,mainly the file "checksum" without the need to download the file to compare,with is mandatory to be done in both scenarios i've described

**Comment by Lonami** [comment]
* The IP layer has a checksum.
* The TCP layer has a checksum.
* [The default connection mode has a checksum](https://github.com/LonamiWebs/Telethon/blob/ee0fc5cc29ad6c7cabbd36edeb5e0355dd98b019/telethon/network/connection/tcpfull.py#L32).
* [The decrypted data is hashed and has to match another checksum](https://github.com/LonamiWebs/Telethon/blob/ee0fc5cc29ad6c7cabbd36edeb5e0355dd98b019/telethon/network/mtprotostate.py#L136).
* The decrypted packet has to make sense or decoding it will fail.

It is *extremely* unlikely that **all** of these fail at once. It's probably more likely that Telegram reads wrong data from disk and sends that.

If you need additional checks, I think you should be the one to do those, not the library. The library helps you upload and download files (and it has enough checksums already). Anything else is up to you.

**Comment by mzramna** [comment]
this is just to the sending way,the reciving way is not here,becouse machines could fail becouse of power issues anthe others,
if anyhow my machine stops the file download in half,the lib cannot detect if the files have been downloaded correctly or not,ne only way that i have to prevent this is this:
```
    if type(message.media) is telethon.tl.types.MessageMediaDocument:
        for i in message.media.document.attributes:
            if type(i) is telethon.tl.types.DocumentAttributeFilename:
                file=i.file_name
                if not os.path.isfile(file):
                    client.download_media(message,progress_callback=callback,file="./"+file+".tmp")
                    os.rename("./"+file+".tmp","./"+file)
```
but this is way less efficient than an official way
btw,this method seems to works in mostly of my tests, but if anyhow the files got renamed it stops working

**Comment by Lonami** [comment]
Okay, downloading to a temporary path and renaming it to the right location iff download completes is far more reasonable than trying to keep a checksum somewhere else. That can be done.

**Comment by Lonami** [comment]
What about people relying on accessing the file while the download is ongoing, though? See https://t.me/TelethonUpdates/1122.

**Comment by Lonami** [comment]
Another issue is having the temporary directory be in a different storage or partition (https://t.me/TelethonChat/280970), then the final step will have to be a copy.

**Comment by Lonami** [comment]
>  cannot continue the previous download

You can do that by specifying an offset with `iter_download`.

**Comment by Lonami** [comment]
Given the poll results, even though I didn't really get much of an explanation for yes, it might be surprising to make `download_media` not create the file as soon as it needs to. This would also cause issues if it turns out the file is not writable, because now the entire download has to complete before we realize.

> When using client.download_media to download media into a local path (str or Path), do you use the partial contents of the file as it gets downloaded in some way? That is, do you read or stat the same path while the download is ongoing?
> 
> ```
> 33% (27 votes) Yes ======
> 67% (55 votes)  No =============
> 82 votes
> ```

I think this should be your own method:

```python
async def full_download(client, media, output):
    import tempfile
    import os
    with tempfile.TemporaryDirectory() as dir:
        name = await client.download_media(media, dir)
        os.rename(name, os.path.join(output, name))
```

(You can easily adapt this to use a specific file name if you don't want the autogenerated name).
